KAFKA-15853: Move consumer group and group coordinator configs out of core

OmniaGM opened the pull request:
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

OmniaGM committed:
KAFKA-15853: Move consumer group configs out of core

chia7712 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorConfig.java:
```diff
@@ -17,13 +17,84 @@
 package org.apache.kafka.coordinator.group;
 
 import org.apache.kafka.coordinator.group.assignor.PartitionAssignor;
+import org.apache.kafka.common.utils.Utils;
```
It seems we don't have strict import orders, but it would be nice to optimize the imports :)

chia7712 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorConfig.java:
```diff
@@ -17,13 +17,84 @@
+
+    public final static String CONSUMER_GROUP_ASSIGNORS_CONFIG = "group.consumer.assignors";
+    public final static String CONSUMER_GROUP_ASSIGNORS_DOC = "The server side assignors as a list of full class names. The first one in the list is considered as the default assignor to be used in the case where the consumer does not specify an assignor.";
+    public static final List<String> CONSUMER_GROUP_ASSIGNORS_DEFAULT = Arrays.asList(
```
Could we make it immutable?

chia7712 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetConfig.java:
```diff
@@ -88,10 +125,10 @@ public OffsetConfig(int maxMetadataSize,
     }
 
     public OffsetConfig() {
```
it seems this constructor is unused. Could we remove it?

OmniaGM replied:
deleted

chia7712 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetConfig.java:
```diff
@@ -20,16 +20,53 @@
+    public static final short OFFSET_COMMIT_REQUIRED_ACKS_DEFAULT = -1;
+    public static final String OFFSET_COMMIT_REQUIRED_ACKS_DOC = "The required acks before the commit can be accepted. In general, the default (-1) should not be overridden.";
+
+    public static final long OFFSET_RETENTION_MS_DEFAULT = 24 * 60 * 60 * 1000L;
```
this default value is not equal to `OFFSETS_RETENTION_MINUTES_DEFAULT`. that is weird to me ...

OmniaGM replied:
yeah this is weird also  because `OFFSET_RETENTION_MS_DEFAULT` has only be used as a default value for `OffsetConfig` constructor in the original scala code https://github.com/apache/kafka/pull/15161/files#diff-157be940464ab3026e373cbc03b515adc52cdc315aba96ecf384e1db85a93f29. While `GroupCoordinator::offsetConfig` was setting this value to `config.offsetsRetentionMinutes * 60L * 1000L`.

I'll delete it as seems like no one is actually referring to this default value

chia7712 reviewed (commented):
@OmniaGM thanks for refactoring code base! a couple of comments left.

OmniaGM committed:
remove unused constructor

OmniaGM committed:
make CONSUMER_GROUP_ASSIGNORS_DEFAULT immutable

OmniaGM committed:
remove OFFSET_RETENTION_MS_DEFAULT
- value has no usage
- and it is the wrong value as it is not equal the default retention in minutes

OmniaGM committed:
fix import order

chia7712 commented:
@OmniaGM it seems there are build error. could you please check them? thanks

OmniaGM committed:
fix pipeline

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetConfig.java:
```diff
@@ -20,16 +20,51 @@
 import org.apache.kafka.common.record.CompressionType;
 
 public class OffsetConfig {
```
@OmniaGM Thanks for working on this. I wonder whether we could merge `OffsetConfig` into `GroupCoordinatorConfig` in order to have one config for the entire package. It would bring us closer to the goal described in https://issues.apache.org/jira/browse/KAFKA-15089. Would it be possible?

OmniaGM replied:
Good call, this actually will make it simplier. Will update the PR to reflict this

OmniaGM replied:
I'll move all props, defaults and docs into `GroupCoordinatorConfig` and move the constructor for `OffsetConfig` as static class within `GroupCoordinatorConfig` as `GroupMetadataManager` which mainly cares about `OffsetConfig` doesn't need the rest of the configs in `GroupCoordinatorConfig`

OmniaGM replied:
I pushed the changes let me know if it makes sense or if you have other suggestions/feedback.

chia7712 replied:
it seems to me the package `GroupCoordinatorConfig.OffsetConfig` is a bit confused, and move `OffsetConfig` out of `GroupCoordinatorConfig` should be fine. WDYT?

OmniaGM replied:
It is also bit odd to have a class  only for the constructor that do nothing but grouping. But I can see that split it out might be better as it seems like we have `GroupConfig` in scala that will get migrated to java at some point so wouldn't make sense to have the `GroupCoordinatorConfig` grow out of hand.

dajac replied:
`OffsetConfig` is only used by Scala code so it will disappear when we remove it. The `GroupConfig` in Scala may never be migrated to Java as `GroupCoordinatorConfig` already contains everything, I think. I wonder if we could actually replace `OffsetConfig` by an interface and make `GroupCoordinatorConfig` implements it. Then, we could pass `GroupCoordinatorConfig` to the old code too. I am not sure if this is feasible though.

chia7712 replied:
> OffsetConfig is only used by Scala code so it will disappear when we remove it

or we can move `OffsetConfig` to `kafka.coordinator.group` since it is used by `kafka.coordinator.group.GroupCoordinator` and `kafka.coordinator.group.GroupMetadataManager`.

OmniaGM replied:
I moved it to `kafka.coordinator.group` for now we can have a followup for
[KAFKA-15089](https://github.com/apache/kafka/pull/15684/commits/581e9a88089aa9881ec28de75877350e17711b3e) to look into removing this class later.

OmniaGM replied:
I would suggest that we wait until KafkaConfig is fully migrated out of core and then we can change the constructor of `GroupCoordinatorConfig` to accept [KafkaConfig](https://issues.apache.org/jira/browse/KAFKA-15853?page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel&focusedCommentId=17789703#comment-17789703) and it extract any needed grouping out KafkaConfig definition. WDYT? This might be the easiest way

chia7712 replied:
> then we can change the constructor of GroupCoordinatorConfig to accept KafkaConfig 

that makes sense to me. A unified way to generate those config class can make consistent behavior.

> I moved it to kafka.coordinator.group 

I assume the package you mentioned is in core module, but I'm ok with your approach since it can avoid rewriting java code back to scala code.

OmniaGM replied:
sorry I meant `org.apache.kafka.coordinator.group`

OmniaGM replied:
I updated the pr now. And rebased from trunk as well

OmniaGM committed:
KAFKA-15089: merge OffsetConfig and GroupCoordinatorConfig

OmniaGM committed:
fix rebase

chia7712 commented:
@OmniaGM it seems there are build error. could you fix them?

OmniaGM committed:
fix pipeline

OmniaGM commented:
ugh my local environment keep the wrong caches because I keep jump between branches

chia7712 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorConfig.java:
```diff
@@ -16,14 +16,132 @@
+    public static final int CONSUMER_GROUP_MAX_HEARTBEAT_INTERVAL_MS_DEFAULT = 15000;
+
+    public final static String CONSUMER_GROUP_MAX_SIZE_CONFIG = "group.consumer.max.size";
+    public final static String CONSUMER_GROUP_MAX_SIZE_DOC = "The maximum number of consumers that a single consumer group can accommodate.";
```
this is unrelated to this PR, but this doc is equal to `GROUP_MAX_SIZE_DOC` (classic version). It would be nice to make some difference for them.

OmniaGM replied:
I believe most of the docs are in more or less similar situation here. For example `CONSUMER_GROUP_MIN_SESSION_TIMEOUT_MS_DOC` is the first statement of `GROUP_MIN_SESSION_TIMEOUT_MS_DOC`. I pushed a small update for `CONSUMER_GROUP_MAX_SIZE_DOC` let me know if it make sense now

chia7712 replied:
the fix is good to me. We can update those docks of new consumer configs in the future.

chia7712 reviewed (approved):
@OmniaGM thanks for doing this refactor. a unrelated comment left. we can address it in another PR

OmniaGM committed:
update CONSUMER_GROUP_MAX_SIZE_DOC

chia7712 commented:
@OmniaGM Could you fix the conflicts?

chia7712 commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -902,40 +824,40 @@ object KafkaConfig {
+      .define(GroupCoordinatorConfig.OFFSETS_TOPIC_REPLICATION_FACTOR_CONFIG, SHORT, GroupCoordinatorConfig.OFFSETS_TOPIC_REPLICATION_FACTOR_DEFAULT, atLeast(1), HIGH, GroupCoordinatorConfig.OFFSETS_TOPIC_REPLICATION_FACTOR_DOC)
+      .define(GroupCoordinatorConfig.OFFSETS_TOPIC_PARTITIONS_CONFIG, INT, GroupCoordinatorConfig.OFFSETS_TOPIC_PARTITIONS_DEFAULT, atLeast(1), HIGH, GroupCoordinatorConfig.OFFSETS_TOPIC_PARTITIONS_DOC)
+      .define(GroupCoordinatorConfig.OFFSETS_TOPIC_SEGMENT_BYTES_CONFIG, INT, GroupCoordinatorConfig.OFFSETS_TOPIC_SEGMENT_BYTES_DEFAULT, atLeast(1), HIGH, GroupCoordinatorConfig.OFFSETS_TOPIC_SEGMENT_BYTES_DOC)
+      .define(GroupCoordinatorConfig.OFFSETS_TOPIC_COMPRESSION_CODEC_CONFIG, INT, GroupCoordinatorConfig.OFFSETS_TOPIC_COMPRESSION_CODEC_DEFAULT, HIGH, GroupCoordinatorConfig.OFFSETS_TOPIC_COMPRESSION_CODEC_DOC)
```
we need to pass `GroupCoordinatorConfig.OFFSETS_TOPIC_COMPRESSION_CODEC_DEFAULT.id.toInt`.

OmniaGM replied:
good spot. Fixed now

OmniaGM committed:
fix OFFSETS_TOPIC_COMPRESSION_CODEC default value

OmniaGM committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-15853-KafkaConfig-prop-consumer-group-coordinator

chia7712 commented:
@OmniaGM Could you please fix those conflicts? thanks!

OmniaGM committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-15853-KafkaConfig-prop-consumer-group-coordinator

chia7712 reviewed (approved):
LGTM
